\documentclass[11pt,notitlepage]{article}
\usepackage{graphicx}
\usepackage{amsmath}            % adds more math symbols
\usepackage{amssymb}
\usepackage{multicol}
\title{Neisseria Ortholog Analysis Tool}
\author{Leo Przybylski\\
\texttt{przybyls@arizona.edu}}

\newcommand{\question}[2]{\textbf{#1.} #2}
\newcommand{\subquestion}[2]{\par\hspace{0.5cm} \textbf{#1)} #2}
\newenvironment{answer}{\endpar%

}

    % Give wider margins; gives more text per page.

\setlength{\topmargin}{0.00in}
\setlength{\textheight}{8.75in}
\setlength{\textwidth}{6.625in}
\setlength{\oddsidemargin}{0.0in}
\setlength{\evensidemargin}{0.0in}

\setlength{\parindent}{0.0cm}	% Don't indent the paragraphs
%\setlength{\parskip}{0.4cm}	% distance between paragraphs

\begin{document}
  \maketitle
  \tableofcontents

  \abstract{\ldots
}
  {\setlength{\baselineskip}%
           {0.0\baselineskip}
  \section*{\hfill Background}
  \hrulefill \par}
  When a blast query is made, results are recorded with the identifier of the query
  as well as the identifier for the result or hit. The convention used to name the
  each query is a combination of gene and contig number. The records of blast output
  can be organized into several different types of meaningful information.


 %% \includegraphics[width=120mm]{api/images/sample-graph1.png}

  \includegraphics[width=145mm]{i90_a90_graph.png}


  One way in particular is to group records related by query and or hit information as
  clusters. We can then provide statistical information on clusters. By grouping clusters
  with the same cardinality of edges, we can analyze trends between different families
  of bacteria. One can analyze that similar families can grow at the same rate or that 
  similar families can recombine with similar genes. This is the benefit of ortholog
  analysis of gene clusters.

  {\setlength{\baselineskip}%
           {0.0\baselineskip}
  \section*{\hfill Clusters}
  \hrulefill \par}
  A Cluster is a set of edges where the query identifiers match. For example, in the
  following blast results:

  \noindent \verb|AP206_contig00001_4923-1612	cinerea_contig00013_10696-7277	47.65	1129	517	23	29	1103	31	1139	0.0	 884|

  \noindent \verb|AP206_contig00001_4923-1612	elongata_contig01464_47682-51131	46.11	1156	545	24	7	1103	13	1149	0.0	 851|
  \\

  The records belong in the same cluster because they were results of the same query. Not
  all records that match the query identifier become edges. 

  \subsection{Identifying Cluster Edges} 
  
  A gene may relate to another gene cluster by defining a cross-cluster relationship. For example, if we added the 
  following data:

  \noindent \verb|sicca_contig00265_4037-6052	AP206_contig00001_4923-1612	24.92	309	170	14	408	671	812	1103	4e-04	42.4|

  The above will create an edge between \emph{AP206\_contig00001\_4923\-1612} and \emph{sicca\_contig00265\_4037\-6052}
  clusters. Then, all of the hits relating to the \emph{sicca\_contig00265\_4037\-6052} query will
  suddenly be added to the cluster as well.

  The edge must also fulfill 
  identity percentage and alignment length ratio requirements. The identity percentage must
  be with in the bounds specified at runtime. The identity percentage requirement is within the range 
  $\in \{30, 45, 60, 75, 90\}$. Likewise, the alignment length ratio must be within the range
  $\in \{50, 70, 90\}$ which is defined at runtime. Again, the identity percentage and
  alignment length ratio requirements are variable and determined at runtime. This study
  provided results on each permutation of identity percentage and alignment length ratio.



  {\setlength{\baselineskip}%
           {0.0\baselineskip}
  \section*{\hfill Conclusion}
  \hrulefill \par}
  
  \subsection{Poisson Distributions}

  \includegraphics[width=145mm]{i30_a50_graph.png}
  
  \includegraphics[width=145mm]{i30_a70_graph.png}
  
  \includegraphics[width=145mm]{i30_a90_graph.png}
  
  \includegraphics[width=145mm]{i45_a50_graph.png}
  
  \includegraphics[width=145mm]{i45_a70_graph.png}
  
  \includegraphics[width=145mm]{i45_a90_graph.png}
  
  \includegraphics[width=145mm]{i60_a50_graph.png}
  
  \includegraphics[width=145mm]{i60_a70_graph.png}
  
  \includegraphics[width=145mm]{i60_a90_graph.png}
  
  \includegraphics[width=145mm]{i75_a50_graph.png}
  
  \includegraphics[width=145mm]{i75_a70_graph.png}
  
  \includegraphics[width=145mm]{i75_a90_graph.png}
  
  \includegraphics[width=145mm]{i90_a50_graph.png}
  
  \includegraphics[width=145mm]{i90_a70_graph.png}
  
  \includegraphics[width=145mm]{i90_a90_graph.png}



  {\setlength{\baselineskip}%
           {0.0\baselineskip}
  \section*{\hfill Classes}
  \hrulefill \par}

  \includegraphics[width=120mm]{api/images/blast_clusters1_Class1.png}  

  \include{api/BlastParser}
  \include{api/BlastRecordHandler}
  \include{api/BlastRecord}
  \include{api/ClusterMatrix}
  \include{api/Cluster}
  \include{api/CommentHandler}
  \include{api/Edge}
  \include{api/RecordHandler}

\end{document}

